scripts/runnable_recipe_ray_vllm_weight_sync.py

```diff
@@ -1582,6 +2031,204 @@ def close(self):
+
+    def acquire_state_dict_lock(self):
+        self.state_dict_lock.acquire_write()
+    
+    def release_state_dict_lock(self):
+        self.version += 1
+        self.version_tensor += 1
+        torch.cuda.synchronize()
+        self.state_dict_lock.release_write()
+    
+    def all_worker_ids(self):
+        return [i for i in range(len(self.collector._remote_collectors))]
+        
+    def _get_server_weights(self):
+        return self.state_dict
+    
+    def _maybe_map_weights(self, server_weights):
+        return server_weights
+    
+    # def _skip_update(self, worker_id):
+    #     if self.version == 0:
+    #         return True
+    #     if worker_id not in self.vllm_weight_versions:
+    #         return False
+    #     if self.vllm_weight_versions[worker_id] == self.version:
+    #         print(f"skipping update for {worker_id=}, {self.version=}, {self.vllm_weight_versions[worker_id]=}")
+    #         return True
+    #     return False
+
+    def _skip_update(self, worker_id):
+        if self.version == 0:
+            return True
+        return False
+    
+    def _init_model_update_group(self, worker_id):
+        # here again, I want to grab the tp size from the vLLM worker... :(
+        # llm.llm_engine.parallel_config.tensor_parallel_size
+        vllm_tp_size = 1
+        weight_sync_world_size = vllm_tp_size + 1
+        print("before stateless_init_process_group")
+        model_update_group = stateless_init_process_group(
+            self.vllm_master_addresses[worker_id],
+            self.vllm_master_ports[worker_id],
+            0,
+            weight_sync_world_size,
+            torch.device("cuda:0"),
+        )
+        print("after stateless_init_process_group")
+        self.vllm_comm_groups[worker_id] = model_update_group
+
+    def _sync_weights_with_worker(
+        self, worker_id: int, server_weights
+    ):
+        print(f"in _sync_weights_with_worker {worker_id}", flush=True)
+        # self.collector._remote_collectors[worker_id].update_policy_weights_.remote()
+        if worker_id not in self.vllm_comm_groups:
+            print("init model update group")
+            self._init_model_update_group(worker_id)
+        print("done init model update group")
+        self.state_dict_lock.acquire_read()
+        for i, k in enumerate(server_weights.keys()):
+            # if i == 0:
+            #     print(f"{server_weights[k][0]=}")
+            self.vllm_comm_groups[worker_id].broadcast(server_weights[k], src=0, stream=torch.cuda.current_stream())
+        self.vllm_comm_groups[worker_id].broadcast(self.version_tensor, src=0, stream=torch.cuda.current_stream())
+        torch.cuda.synchronize()
+        print(f"_sync_weights_with_worker done broadcast {worker_id} {self.version=}")
+        self.vllm_weight_versions[worker_id] = self.version
+        self.state_dict_lock.release_read()
+
+
+@ray.remote(num_cpus=4, num_gpus=1)
+class vLLMParameterServer(vLLMRemoteWeightUpdaterBase):
```

Comment:
Do we need this class?

Reply:
If you don't have this class in this script where are you sending the trainer weights to? 

This is an example of what a param server could look like but is by no means the gold standard indeed

There are other alternatives eg dcp.save and dcp.load where param server just manages versioning etc.

But this is one of the efficiency vs runnable things where we discussed the efficiency/method of the weight sync is not our priority for now, no?

Reply:
No I meant: do we need to subclass given we already have a `vLLMRemoteWeightUpdater`?